coderd/updatecheck/updatecheck.go

```diff
@@ -0,0 +1,229 @@
+	// default 24h.
+	Interval time.Duration
+	// Notify is called when a newer version of Coder (than the
+	// last update check) is available.
+	Notify func(r Result)
+}
+
+// New returns a new Checker that periodically checks for Coder updates.
+func New(db database.Store, log slog.Logger, opts Options) *Checker {
+	if opts.Client == nil {
+		opts.Client = http.DefaultClient
+	}
+	if opts.URL == "" {
+		opts.URL = defaultURL
+	}
+	if opts.Interval == 0 {
+		opts.Interval = 24 * time.Hour
+	}
+	if opts.Notify == nil {
+		opts.Notify = func(r Result) {}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &Checker{
+		ctx:        ctx,
+		cancel:     cancel,
+		db:         db,
+		log:        log,
+		opts:       opts,
+		firstCheck: make(chan struct{}),
+		closed:     make(chan struct{}),
+	}
+	go c.start()
+	return c
+}
+
+// Result is the result from the last update check.
+type Result struct {
+	Checked time.Time `json:"checked,omitempty"`
+	Version string    `json:"version,omitempty"`
+	URL     string    `json:"url,omitempty"`
+}
+
+// Latest returns the latest version of Coder.
+func (c *Checker) Latest(ctx context.Context) (r Result, err error) {
+	select {
+	case <-c.ctx.Done():
+		return r, c.ctx.Err()
+	case <-ctx.Done():
+		return r, ctx.Err()
+	case <-c.firstCheck:
+	}
+
+	return c.lastUpdateCheck(ctx)
+}
+
+func (c *Checker) init() (Result, error) {
+	defer close(c.firstCheck)
+
+	r, err := c.lastUpdateCheck(c.ctx)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return Result{}, xerrors.Errorf("last update check: %w", err)
+	}
+	if r.Checked.IsZero() || time.Since(r.Checked) > c.opts.Interval {
+		r, err = c.update()
+		if err != nil {
+			return Result{}, xerrors.Errorf("update check failed: %w", err)
+		}
+	}
+
+	return r, nil
+}
+
+func (c *Checker) start() {
+	defer close(c.closed)
+
+	r, err := c.init()
+	if err != nil {
+		if xerrors.Is(err, context.Canceled) {
+			return
+		}
+		c.log.Error(c.ctx, "init failed", slog.Error(err))
+	} else {
+		c.opts.Notify(r)
+	}
+
+	t := time.NewTicker(c.opts.Interval)
+	defer t.Stop()
+
+	diff := time.Until(r.Checked.Add(c.opts.Interval))
+	if diff > 0 {
+		c.log.Info(c.ctx, "time until next update check", slog.F("duration", diff))
+		t.Reset(diff)
+	} else {
+		c.log.Info(c.ctx, "time until next update check", slog.F("duration", c.opts.Interval))
+	}
+
+	for {
+		select {
+		case <-t.C:
+			rr, err := c.update()
+			if err != nil {
+				if xerrors.Is(err, context.Canceled) {
+					return
+				}
+				c.log.Error(c.ctx, "update check failed", slog.Error(err))
+			} else {
+				c.notifyIfNewer(r, rr)
+				r = rr
+			}
+			c.log.Info(c.ctx, "time until next update check", slog.F("duration", c.opts.Interval))
+			t.Reset(c.opts.Interval)
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Checker) update() (r Result, err error) {
+	c.log.Info(c.ctx, "checking for update")
+	req, err := http.NewRequestWithContext(c.ctx, http.MethodGet, c.opts.URL, nil)
```

Comment:
Should use a new context with a low timeout, maybe 5 seconds or something.

Reply:
Great suggestion, I'll go with 30 secs since this is a non-critical service and the check only runs 1/day.

Reply:
30 seconds seems very high for a single API request, I can't imagine the GitHub API ever taking more than a few seconds to respond and the latency to the server being higher than a couple of seconds unless GitHub is down, which you're gonna get a bad gateway res anyways

Reply:
There are many failure scenarios, sometimes it could be local network hickups or the GitHub service could be partially working (slowly).

I'd like to understand why 30s would be an issue in this usage scenario? Unless we implement a retry-mechanism I don't see any benefit to making it shorter.

Reply:
There already is a retry mechanism, it'll retry in 24 hours 😄 

30 seconds is probably fine though.